### DISCUSSION TITLE
Berkeley SPICE Compatibility

### DISCUSSION THREAD

**Discussion by 4b11b4** [SpiceSharp/SpiceSharp#168] [General]
Hi there,

This looks to be an excellent project!

I was just curious if you might elaborate on how this library is 1:1 compatible with SPICE (3f5).
Did you essentially reverse engineer SPICE?

Thanks for your time!
Anton

**Comment by svenboulanger** [comment]
Short answer: yes.
This was also not the first time I attempted to reverse-engineer Spice 3f5 of course. I am just an electronics engineer with a passion for numerical algorithms as well.

The 1:1 compatibility was originally achieved by extracting simulation data from Spice 3f5 and comparing it to Spice#. In some cases, I even stepped through both Spice 3f5 and Spice# side-by-side to find where it goes wrong. I started to discover some "issues" with Spice 3f5 that didn't fully make sense. For example, 3f5 exported data of the second-last iteration rather than the last (most accurate) solution. Also with transient simulations, comparing is very difficult as even the smallest error can cause wildly different timesteps, which can easily accumulate over time.

So I do keep the reference simulation data of Spice 3f5 in the unit tests for many models, guaranteeing a 1:1 relationship. We kind of have passed that point though, as I have added stuff that don't really have a counterpart in Spice 3f5.
